agent/php_user_instrument_wraprec_hashmap.c

```diff
@@ -0,0 +1,485 @@
+  hashmap->elements = 0;
+
+  return hashmap;
+}
+
+static size_t nr_func_hashmap_hash_key(size_t log2_num_buckets, nr_func_hashmap_key_t* key) {
+  return (key->name_hash & ((1 << log2_num_buckets) - 1));
+}
+
+static bool nr_func_hashmap_key_equals(nr_func_hashmap_key_t* a, nr_func_hashmap_key_t* b) {
+  return (a->name_hash == b->name_hash) && (a->name_len == b->name_len)
+         && (0 == nr_strncmp(a->name, b->name, a->name_len));
+}
+
+static bool nr_func_hashmap_fetch_internal(nr_func_hashmap_t* hashmap, size_t hash, nr_func_hashmap_key_t* key, nr_func_bucket_t** bucket_ptr) {
+  nr_func_bucket_t* bucket;
+  for (bucket = hashmap->buckets[hash]; bucket; bucket = bucket->next) {
+    if (nr_func_hashmap_key_equals(bucket->key, key)) {
+      if (bucket_ptr) {
+        *bucket_ptr = bucket;
+      }
+      return true;
+    }
+  }
+  return false;
+}
+
+static nruserfn_t* nr_func_hashmap_add_internal(nr_func_hashmap_t* hashmap,
+                             size_t hash_key,
+                             nr_func_hashmap_key_t* key) {
+  nr_func_bucket_t* bucket;
+
+  bucket = (nr_func_bucket_t*)nr_malloc(sizeof(nr_func_bucket_t));
+  bucket->prev = NULL;
+  bucket->next = hashmap->buckets[hash_key];
+  bucket->key = (nr_func_hashmap_key_t*)nr_malloc(sizeof(nr_func_hashmap_key_t));
+  bucket->key->is_method = key->is_method;
+  bucket->key->name = nr_strndup(key->name, key->name_len);
+  bucket->key->name_len = key->name_len;
+  bucket->key->name_hash = key->name_hash;
+  bucket->wraprec = nr_php_user_wraprec_create();
+
+  if (hashmap->buckets[hash_key]) {
+    hashmap->buckets[hash_key]->prev = bucket;
+  }
+
+  hashmap->buckets[hash_key] = bucket;
+  ++hashmap->elements;
+
+  return bucket->wraprec;
+}
+
+static nruserfn_t* nr_func_hashmap_lookup_internal(nr_func_hashmap_t* hashmap, nr_func_hashmap_key_t* key) {
+  size_t hash;
+  nr_func_bucket_t* bucket;
+
+  if (nrunlikely((NULL == hashmap) || (NULL == key))) {
+    return NULL;
+  }
+
+  hash = nr_func_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_func_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    return bucket->wraprec;
+  }
+
+  return NULL;
+}
+
+static nruserfn_t* nr_func_hashmap_update_internal(nr_func_hashmap_t* hashmap, nr_func_hashmap_key_t* key, bool* created) {
+  size_t hash;
+  nr_func_bucket_t* bucket;
+  nruserfn_t* wraprec;
+
+  if (nrunlikely((NULL == hashmap) || (NULL == key))) {
+    return NULL;
+  }
+
+  hash = nr_func_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_func_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    if (created) {
+      *created = false;
+    }
+    return bucket->wraprec;
+  }
+
+  if (created) {
+    *created = true;
+  }
+  return nr_func_hashmap_add_internal(hashmap, hash, key);
+}
+
+static void nr_func_hashmap_destroy_bucket_internal(nr_func_bucket_t** bucket_ptr) {
+  nr_func_bucket_t* bucket = *bucket_ptr;
+  nr_free(bucket->key->name);
```

Comment:
Possibly good idea to NULL check these struct pointer usage?

Reply:
This function is used as a helper and in that context `*bucket_ptr` is guaranteed to be non-NULL.  However, I believe a NULL check for `bucket->key` is warranted as that field is not as readily obvious to not possibly be NULL.

Reply:
Added in c9563c2c11298078e426c88b6c39380812e5091f.